### ISSUE TITLE
Panic when calling encode_update_v2 on TransactionMut

### ISSUE THREAD

**Issue by stellit** [y-crdt/y-crdt#295]
Hi!

There is a mysterious behavior found with the TransactionMut.encode_update_v2() method.

In the given test code, if line `[A]` with the first `encode_update_v2` commented out, the code will crashes at the line `[B]`. But if enable the call `[A]`, `[B]` will pass smoothly. I've also attached the same logic that tested in yjs:

Here is the code in rust:
```rust
#[test]
fn test_yrs_issue() {
    let before = "AAEAC4HX78kQxu2Y3AgEBAAEAQEACygDJwAEAAcAKAMIowGOAW5vdGUuZ3VpZG5vdGVHdWlkbm90ZS5vd25lcm93bmVybm90ZS50eXBlbm90ZVR5cGVub3RlLmNyZWF0ZVRpbWVjcmVhdGVUaW1lbm90ZS50aXRsZXRpdGxlMXByb3NlbWlycm9ybGlua25vdGVndWlkc2xvdEd1aWR0eXBlbGlua1R5cGVjaGlsZHJlbmIJCAoFCQgPSgAFAQsIBAgESAABCQEEAAABAAADAQICAwJBCAIEAHcWYWNGeFVZRExSaGVya0phQmVjc2MzZ32IOX0Adw0xNjg1NTMwMjg2NjU2CQB3FkJDZFFwcHdFVDBpUlZCUS04RVcyV2d3FmpyRW1JTXBWVGUtY3JOMlZMM2NhSFF3CGxpbmtub3RldwExdgIEZ3VpZHcWQkNkUXBwd0VUMGlSVkJRLThFVzJXZwhzbG90R3VpZHcWanJFbUlNcFZUZS1jck4yVkwzY2FIUXcWQkNkUXBwd0VUMGlSVkJRLThFVzJXZwA";
    let update = "AAAAAAAAAQAAAAAAAca2jK4EAQIG";
    let raw_doc = decode_base64_url_safe(before).unwrap();
    let update = decode_base64_url_safe(update).unwrap();
    let doc = empty_doc(true);
    {
        let mut txn = doc.transact_mut();
        txn.apply_update(Update::decode_v2(&raw_doc).unwrap());
        // txn.encode_update_v2(); // [A] first encode_update_v2() call
    }
    {
        let b = doc.get_or_insert_array("b");
        let txn = doc.transact();
        println!("before update: {}", b.to_json(&txn));
    }
    {
        let mut txn = doc.transact_mut();
        txn.apply_update(Update::decode_v2(&update).unwrap());
        txn.encode_update_v2(); // [B] crashes here when A is commented out
    }
    {
        let b = doc.get_or_insert_array("b");
        let txn = doc.transact();
        println!("after update: {}", b.to_json(&txn));
    }
}
```

When the line `[A]` is commented out, the code will crashes at `encode.rs#write_ds_clock:L269`. It seems the clock is over reduced during the encoding process.

Here are the helper method that used to create empty doc as well as decode the base64url encoded string.
```rust
fn empty_doc(skip_gc: bool) -> Doc {
    Doc::with_options(Options {
        skip_gc,
        ..Default::default()
    })
}

fn decode_base64_url_safe(input: &str) -> Vec<u8> {
    match base64::decode::decode_config(input, base64::Config::new(base64::CharacterSet::UrlSafe, false)) {
        Ok(v) => Ok(v),
        Err(e) => Ok(vec![]),
    }
}
```

Here is the test code in yjs:
```typescript
test('test malform', () => {
    const doc = init(false);
    const before = "AAEAC4HX78kQxu2Y3AgEBAAEAQEACygDJwAEAAcAKAMIowGOAW5vdGUuZ3VpZG5vdGVHdWlkbm90ZS5vd25lcm93bmVybm90ZS50eXBlbm90ZVR5cGVub3RlLmNyZWF0ZVRpbWVjcmVhdGVUaW1lbm90ZS50aXRsZXRpdGxlMXByb3NlbWlycm9ybGlua25vdGVndWlkc2xvdEd1aWR0eXBlbGlua1R5cGVjaGlsZHJlbmIJCAoFCQgPSgAFAQsIBAgESAABCQEEAAABAAADAQICAwJBCAIEAHcWYWNGeFVZRExSaGVya0phQmVjc2MzZ32IOX0Adw0xNjg1NTMwMjg2NjU2CQB3FkJDZFFwcHdFVDBpUlZCUS04RVcyV2d3FmpyRW1JTXBWVGUtY3JOMlZMM2NhSFF3CGxpbmtub3RldwExdgIEZ3VpZHcWQkNkUXBwd0VUMGlSVkJRLThFVzJXZwhzbG90R3VpZHcWanJFbUlNcFZUZS1jck4yVkwzY2FIUXcWQkNkUXBwd0VUMGlSVkJRLThFVzJXZwA";
    const update = "AAAAAAAAAQAAAAAAAca2jK4EAQIG";

    doc.on('updateV2', (updates: Uint8Array) => {
        console.debug(`got an onUpdate v2 event.`);
        console.debug(`updates: ${Buffer.from(updates).toString('base64url')}`);
    });


    Y.applyUpdateV2(doc, Buffer.from(before, 'base64url'));
    console.debug(`doc after initial content ${JSON.stringify(doc.getArray('b'))}`);

    Y.applyUpdateV2(doc, Buffer.from(update, 'base64url'));
    console.debug(`doc after update ${JSON.stringify(doc.getArray('b'))}`);

});

const init = (gc: boolean, bin?: Uint8Array): Y.Doc => {
    const tree = new Y.Doc({
        gc,
        gcFilter: () => {
            return gc;
        }
    });
    if (bin && bin.length > 0) {
        Y.applyUpdateV2(tree, bin);
    }
    return tree;
};
```

**Comment by stellit** [comment]
Hi @Horusiath,

do you have an idea why this happens? Is there a easy way to get around of this?

**Comment by Horusiath** [comment]
I'll be working on it tomorrow. For now the easiest work around would be to not use v2 encoding and stick with v1 (which often is a better option anyway).

**Comment by stellit** [comment]
thanks for the advice! Switch to v1 can walk around this issue temporarily.
I've also reduced the test down to the following version:
```rust
#[test]
fn test_v1_v2() {
    let update = "AAHGtoyuBAQCAgYBBAIHAg";
    let update = decode_base64_url_safe(update).unwrap();
    let update_v1 = Update::decode_v1(&update).unwrap();
    let update_v2 = update_v1.encode_v2();
}
```
the encoded base64 string represents the same information as in the initial comment but encoded in v1 format. Hope it helps.

**Comment by stellit** [comment]
I think the issue is cased by the delete set encoding, or let's say the client that produced this update. The range is not really sorted by `start`.
![Screenshot 2023-06-09 at 16 09 50](https://github.com/y-crdt/y-crdt/assets/68103/09f1b971-9809-4a65-8508-24effa7f0460)
As in the screenshot stated, the range is miss ordered by start: `2`, `6`, `4`, `7`
But in yjs, the delete set will be merged into one range instead of 4 ranges: start 2 length 7 after decoded.
The issue could be solved if range can be merged into one or make the range sorted in the Decode implementation for IdRange.

**Comment by Horusiath** [comment]
I've noticed that as well. IdRange compaction is an explicit process and it should happen before serialisation. What worries me, is that the test you've provided, doesn't fail deterministically: depending on the run it sometimes passes, which should not be the case.

**Comment by Horusiath** [comment]
The difference in general is caused by the fact, that in Yjs `DeleteSet` compaction happens prior serialization. In Yrs it's not possible, as serializing update from transaction would require a mutable transaction reference in order to compact delete set (without copying it). Therefore in Yrs `DeleteSet` compaction happens only as part of transaction commit. 

Unfortunately V2 encoding expects `DeleteSet` to be provided in a compacted form and it doesn't check if encoded ID range is not overlapping with previous one.

**Comment by stellit** [comment]
I currently have a `ugly` fix... which is may cause performance issue: I have the range sorted before return, or encoding, which fix the issue temporarily. 
I may misunderstood the "deterministic" you mentioned previously. On the M1 chip, this issue happens consistently. I do not have a chance to test it on a x86 chip.

**Comment by stellit** [comment]
Hi! @Horusiath, it resolved the issue! Thanks!!

**Comment by Horusiath** [comment]
Fix released as part of v0.16.7.
